### PR TITLE
Add iron HTTP server

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

1. Add an `iron` HTTP server with the following endpoints:
    - `/metrics`: Prometheus-format metrics.
    - `/node/fork`: returns the current fork.
1. Creates the `BeaconChainTypes` super-type and threads it through the program.

## Additional Info

Change #1 is the primary goal of this PR and change #2 helped facilitate it.

### CODE REVIEW

File: beacon_node/http_server/src/api.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    fn after(&self, _req: &mut Request, mut resp: Response) -> IronResult<Response> {
+        if resp.headers.get::<ContentType>() == None {
+            resp.headers.set(ContentType::json());
+        }
+        Ok(resp)
+    }
+}
+
+fn handle_fork<T: BeaconChainTypes + 'static>(req: &mut Request) -> IronResult<Response> {
+    // TODO: investigate unwrap - I'm _guessing_ we'll never hit it but we should check to be sure.
```
Comment (michaelsproul) [inline beacon_node/http_server/src/api.rs]: Yeah this should succeed so long as the `persistent::Read` has been linked (which it has)

File: beacon_node/http_server/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    // A `GET` request to `/metrics` is handled by the `metrics` module.
+    router.get(
+        "/metrics",
+        metrics::build_handler(beacon_chain.clone()),
+        "metrics",
+    );
+
+    // Any request to all other endpoints is handled by the `api` module.
+    router.any("/*", api::build_handler(beacon_chain.clone()), "api");
```
Comment (michaelsproul) [inline beacon_node/http_server/src/lib.rs]: Another way to separate these is to "mount" the API handler on `/api/`, and the metrics handler on `/metrics/` using the [`Mount` middleware](https://docs.rs/mount/0.4.0/mount/struct.Mount.html). Then we could have URLs like `/api/node/fork`, etc. Probably not worth changing for now, but would be worth thinking about once we have more APIs and we want to stabilise them

File: beacon_node/http_server/src/api.rs (hunk 1 above)
Comment (paulhauner) [reply beacon_node/http_server/src/api.rs]: Do you think it would be prudent to create a conversion for this error into some `IronResult`?

Comment (paulhauner) [reply beacon_node/http_server/src/api.rs]: Perhaps some sort of `500` result.

Comment (michaelsproul) [reply beacon_node/http_server/src/api.rs]: Yeah, may as well avoid taking the whole node down. Looks like `IronResult::new` takes any `std::Error` and a response

Comment (paulhauner) [reply beacon_node/http_server/src/api.rs]: Fixed!

**Pull request merged**